octobot_trading/orders/order_state.py

```diff
@@ -0,0 +1,184 @@
+
+from octobot_commons.logging.logging_util import get_logger
+
+from octobot_trading.channels.exchange_channel import get_chan
+from octobot_trading.constants import ORDERS_CHANNEL
+from octobot_trading.enums import OrderStates
+from octobot_trading.util.initializable import Initializable
+
+
+class OrderState(Initializable):
+    DEFAULT_SYNC_RETRY_TIMEOUT = 5
+
+    def __init__(self, order, is_from_exchange_data):
+        super().__init__()
+
+        # related order
+        self.order = order
+
+        # default state
+        self.state = OrderStates.UNKNOWN
+
+        # if this state has been created from exchange data or OctoBot internal mechanism
+        self.is_from_exchange_data = is_from_exchange_data
+
+        # the updating task
+        self.updating_task = None
+
+    def is_pending(self) -> bool:
+        """
+        :return: True if the state is pending for update
+        """
+        return self.state is OrderStates.UNKNOWN
+
+    def is_refreshing(self) -> bool:
+        """
+        :return: True if the state is updating
+        """
+        return self.state is OrderStates.REFRESHING
+
+    def is_open(self) -> bool:
+        """
+        :return: True if the Order is considered as open
+        """
+        return not (self.is_filled() or self.is_canceled() or self.is_closed())
+
+    def is_filled(self) -> bool:
+        """
+        :return: True if the Order is considered as filled
+        """
+        return False
+
+    def is_closed(self) -> bool:
+        """
+        :return: True if the Order is considered as closed
+        """
+        return False
+
+    def is_canceled(self) -> bool:
+        """
+        :return: True if the Order is considered as canceled
+        """
+        return False
+
+    def get_logger(self):
+        """
+        :return: the order logger
+        """
+        return get_logger(self.order.get_logger_name())
+
+    def log_order_event_message(self, state_message):
+        """
+        Log an order state event
+        """
+        self.get_logger().info(f"{self.order.symbol} {self.order.get_name()} at {self.order.origin_price}"
+                               f" (ID: {self.order.order_id}) {state_message}"
+                               f" on {self.order.exchange_manager.exchange_name}")
+
+    async def initialize_impl(self) -> None:
+        """
+        Default OrderState initialization process
+        """
+        await self.update()
+
+    async def terminate(self) -> None:
+        """
+        Implement the state ending process
+        Can be portfolio updates, fees request, linked order cancellation, Trade creation etc...
+        """
+        raise NotImplementedError("terminate not implemented")
+
+    async def update(self) -> None:
+        """
+        Update the order state
+        Try to fix the pending state or terminate
+        """
+        if self.is_pending() and self.state is not OrderStates.REFRESHING and self.updating_task is None:
```

Comment:
if self.is_pending() or if **not** self.is_pending() ?

Reply:
**not** :+1:

Reply:
No it's `self.is_pending()`. Pending mean that the Order current state is Unknown. Shouldn't exist.